File: RootTools2/src/main/java/com/rarnu/tools/neo/xposed/SystemCore.java
package com.rarnu.tools.neo.xposed;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.util.Base64;
import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.IXposedHookZygoteInit;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XSharedPreferences;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

public class SystemCore implements IXposedHookZygoteInit, IXposedHookLoadPackage {

    private Context _context = null;
    public void initZygote(IXposedHookZygoteInit.StartupParam paramStartupParam) throws Throwable {
        final XSharedPreferences prefs = new XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF);
        prefs.makeWorldReadable();
        XposedBridge.hookAllMethods(XposedHelpers.findClass("com.android.org.conscrypt.OpenSSLSignature", null), "engineVerify", new XC_MethodHook() {
            protected void beforeHookedMethod(XC_MethodHook.MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                prefs.reload();
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    paramAnonymousMethodHookParam.setResult(true);
                }
            }
        });

        XposedHelpers.findAndHookMethod("java.security.MessageDigest", null, "isEqual", byte[].class, byte[].class, new XC_MethodHook() {
            protected void beforeHookedMethod(MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                prefs.reload();
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    paramAnonymousMethodHookParam.setResult(true);
                }

            }
        });

        XposedHelpers.findAndHookMethod("java.security.Signature", null, "verify", byte[].class, new XC_MethodHook() {
            protected void beforeHookedMethod(MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                prefs.reload();
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    if (((((java.security.Signature) paramAnonymousMethodHookParam.thisObject).getAlgorithm().toLowerCase().equals("sha1withrsa"))
                            || (((java.security.Signature) paramAnonymousMethodHookParam.thisObject).getAlgorithm().toLowerCase().equals("rsa-sha1"))
                            || (((java.security.Signature) paramAnonymousMethodHookParam.thisObject).getAlgorithm().toLowerCase().equals("1.3.14.3.2.26with1.2.840.113549.1.1.1")))
                            &&( XposedHelpers.getIntField(paramAnonymousMethodHookParam.thisObject, "state") == 3)) {
                        paramAnonymousMethodHookParam.setResult(true);
                    }
                }

            }
        });

        XposedHelpers.findAndHookMethod("java.security.Signature", null, "verify", byte[].class, Integer.TYPE, Integer.TYPE, new XC_MethodHook() {
            protected void beforeHookedMethod(MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                prefs.reload();
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    if (((((java.security.Signature) paramAnonymousMethodHookParam.thisObject).getAlgorithm().toLowerCase().equals("sha1withrsa"))
                            || (((java.security.Signature) paramAnonymousMethodHookParam.thisObject).getAlgorithm().toLowerCase().equals("rsa-sha1"))
                            || (((java.security.Signature) paramAnonymousMethodHookParam.thisObject).getAlgorithm().toLowerCase().equals("1.3.14.3.2.26with1.2.840.113549.1.1.1")))
                            && (XposedHelpers.getIntField(paramAnonymousMethodHookParam.thisObject, "state") == 3)) {
                        paramAnonymousMethodHookParam.setResult(true);
                    }
                }
            }
        });
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam paramLoadPackageParam) throws Throwable {
        final XSharedPreferences prefs = new XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF);
        prefs.makeWorldReadable();
        if (paramLoadPackageParam.packageName.equals("android") && paramLoadPackageParam.processName.equals("android")) {
            Class localClass = XposedHelpers.findClass("com.android.server.pm.PackageManagerService", paramLoadPackageParam.classLoader);
            XposedBridge.hookAllConstructors(XposedHelpers.findClass("com.android.server.pm.PackageManagerService", paramLoadPackageParam.classLoader), new XC_MethodHook() {
                protected void afterHookedMethod(XC_MethodHook.MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                    prefs.reload();
                    if (!prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                        return;
                    }
                    _context = ((Context) paramAnonymousMethodHookParam.args[0]);
                }
            });
            XposedBridge.hookAllMethods(localClass, "compareSignatures", new XC_MethodHook() {
                @SuppressWarnings("Duplicates")
                protected void beforeHookedMethod(XC_MethodHook.MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                    prefs.reload();
                    if (!prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                        return;
                    }
                    PackageInfo localObject;
                    int k;
                    int j;
                    android.content.pm.Signature[] arrayOfSignature2;
                    int i;
                    int m;
                    String tmpStr;
                    try {
                        localObject = _context.getPackageManager().getPackageInfo("android", PackageManager.GET_SIGNATURES);
                        if (localObject.signatures[0] == null) {
                            return;
                        }
                        tmpStr = Base64.encodeToString(localObject.signatures[0].toByteArray(), 0).replaceAll("\n", "");
                        k = 0;
                        j = 0;
                        android.content.pm.Signature[] arrayOfSignature1 = (android.content.pm.Signature[]) paramAnonymousMethodHookParam.args[0];
                        arrayOfSignature2 = (android.content.pm.Signature[]) paramAnonymousMethodHookParam.args[1];
                        i = k;
                        if (arrayOfSignature1 != null) {
                            i = k;
                            if (arrayOfSignature1.length > 0) {
                                m = arrayOfSignature1.length;
                                k = 0;
                                for (; ; ) {
                                    i = j;
                                    if (k >= m) {
                                        break;
                                    }
                                    if (Base64.encodeToString(arrayOfSignature1[k].toByteArray(), 0).replaceAll("\n", "").equals(tmpStr)) {
                                        j = 1;
                                    }
                                    k += 1;
                                }
                            }
                        }
                        k = i;
                    } catch (Exception p2) {
                        p2.printStackTrace();
                        return;
                    }
                    if (arrayOfSignature2 != null) {
                        k = i;
                        if (arrayOfSignature2.length > 0) {
                            m = arrayOfSignature2.length;
                            j = 0;
                            for (; ; ) {
                                k = i;
                                if (j >= m) {
                                    break;
                                }
                                if (Base64.encodeToString(arrayOfSignature2[j].toByteArray(), 0).replaceAll("\n", "").equals(tmpStr)) {
                                    i = 1;
                                }
                                j += 1;
                            }
                        }
                    }
                    if (k == 0) {
                        paramAnonymousMethodHookParam.setResult(0);
                    }
                }
            });

        }
        if (paramLoadPackageParam.packageName.equals("com.android.settings")) {
            XposedBridge.hookAllMethods(XposedHelpers.findClass("com.android.settings.applications.AppOpsDetails", paramLoadPackageParam.classLoader), "isPlatformSigned", new XC_MethodHook() {
                protected void beforeHookedMethod(XC_MethodHook.MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                    prefs.reload();
                    if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                        paramAnonymousMethodHookParam.setResult(false);
                    }
                }
            });
        }

    }
}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/xposed/SystemCore.kt
package com.rarnu.tools.neo.xposed

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.util.Base64
import de.robv.android.xposed.IXposedHookLoadPackage
import de.robv.android.xposed.IXposedHookZygoteInit
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XSharedPreferences
import de.robv.android.xposed.XposedBridge
import de.robv.android.xposed.XposedHelpers
import de.robv.android.xposed.callbacks.XC_LoadPackage

class SystemCore : IXposedHookZygoteInit, IXposedHookLoadPackage {

    private var _context: Context? = null

    @Throws(Throwable::class)
    override fun initZygote(paramStartupParam: IXposedHookZygoteInit.StartupParam) {
        val prefs = XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF)
        prefs.makeWorldReadable()
        prefs.reload()


        XposedBridge.hookAllMethods(XposedHelpers.findClass("com.android.org.conscrypt.OpenSSLSignature", null), "engineVerify", object : XC_MethodHook() {
            @Throws(Throwable::class)
            override fun beforeHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                prefs.reload()
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    paramAnonymousMethodHookParam.result = true
                }
            }
        })

        XposedHelpers.findAndHookMethod("java.security.MessageDigest", null, "isEqual", ByteArray::class.java, ByteArray::class.java, object : XC_MethodHook() {
            @Throws(Throwable::class)
            override fun beforeHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                prefs.reload()
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    paramAnonymousMethodHookParam.result = true
                }
            }
        })

        XposedHelpers.findAndHookMethod("java.security.Signature", null, "verify", ByteArray::class.java, object : XC_MethodHook() {
            @Throws(Throwable::class)
            override fun beforeHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                prefs.reload()
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    if (((paramAnonymousMethodHookParam.thisObject as java.security.Signature).algorithm.toLowerCase() == "sha1withrsa"
                            || (paramAnonymousMethodHookParam.thisObject as java.security.Signature).algorithm.toLowerCase() == "rsa-sha1"
                            || (paramAnonymousMethodHookParam.thisObject as java.security.Signature).algorithm.toLowerCase() == "1.3.14.3.2.26with1.2.840.113549.1.1.1")
                            && XposedHelpers.getIntField(paramAnonymousMethodHookParam.thisObject, "state") == 3) {
                        paramAnonymousMethodHookParam.result = true
                    }
                }

            }
        })

        XposedHelpers.findAndHookMethod("java.security.Signature", null, "verify", ByteArray::class.java, Integer.TYPE, Integer.TYPE, object : XC_MethodHook() {
            @Throws(Throwable::class)
            override fun beforeHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                prefs.reload()
                if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                    if (((paramAnonymousMethodHookParam.thisObject as java.security.Signature).algorithm.toLowerCase() == "sha1withrsa"
                            || (paramAnonymousMethodHookParam.thisObject as java.security.Signature).algorithm.toLowerCase() == "rsa-sha1"
                            || (paramAnonymousMethodHookParam.thisObject as java.security.Signature).algorithm.toLowerCase() == "1.3.14.3.2.26with1.2.840.113549.1.1.1")
                            && XposedHelpers.getIntField(paramAnonymousMethodHookParam.thisObject, "state") == 3) {
                        paramAnonymousMethodHookParam.result = true
                    }
                }
            }
        })
    }

    @Throws(Throwable::class)
    override fun handleLoadPackage(paramLoadPackageParam: XC_LoadPackage.LoadPackageParam) {
        val prefs = XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF)
        prefs.makeWorldReadable()
        if (paramLoadPackageParam.packageName == "android" && paramLoadPackageParam.processName == "android") {
            val localClass = XpUtils.findClass(paramLoadPackageParam.classLoader, "com.android.server.pm.PackageManagerService")
            XposedBridge.hookAllConstructors(XposedHelpers.findClass("com.android.server.pm.PackageManagerService", paramLoadPackageParam.classLoader), object : XC_MethodHook() {
                @Throws(Throwable::class)
                override fun afterHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                    prefs.reload()
                    if (!prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                        return
                    }
                    _context = paramAnonymousMethodHookParam.args[0] as Context
                }
            })
            XposedBridge.hookAllMethods(localClass, "compareSignatures", object : XC_MethodHook() {
                @Suppress("UNCHECKED_CAST")
                @SuppressWarnings("Duplicates")
                @Throws(Throwable::class)
                override fun beforeHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                    prefs.reload()
                    if (!prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                        return
                    }
                    val localObject: PackageInfo?
                    var k: Int
                    var j: Int
                    val arrayOfSignature2: Array<android.content.pm.Signature>?
                    var i: Int
                    var m: Int
                    val tmpStr: String
                    try {
                        localObject = _context?.packageManager?.getPackageInfo("android", PackageManager.GET_SIGNATURES)
                        if (localObject?.signatures!![0] == null) {
                            return
                        }
                        tmpStr = Base64.encodeToString(localObject?.signatures!![0].toByteArray(), 0).replace("\n".toRegex(), "")
                        k = 0
                        j = 0
                        val arrayOfSignature1 = paramAnonymousMethodHookParam.args[0] as Array<Signature>?
                        arrayOfSignature2 = paramAnonymousMethodHookParam.args[1] as Array<Signature>?
                        i = k
                        if (arrayOfSignature1 != null) {
                            i = k
                            if (arrayOfSignature1.isNotEmpty()) {
                                m = arrayOfSignature1.size
                                k = 0
                                while (true) {
                                    i = j
                                    if (k >= m) {
                                        break
                                    }
                                    if (Base64.encodeToString(arrayOfSignature1[k].toByteArray(), 0).replace("\n".toRegex(), "") == tmpStr) {
                                        j = 1
                                    }
                                    k += 1
                                }
                            }
                        }
                        k = i
                    } catch (p2: Exception) {
                        p2.printStackTrace()
                        return
                    }

                    if (arrayOfSignature2 != null) {
                        k = i
                        if (arrayOfSignature2.isNotEmpty()) {
                            m = arrayOfSignature2.size
                            j = 0
                            while (true) {
                                k = i
                                if (j >= m) {
                                    break
                                }
                                if (Base64.encodeToString(arrayOfSignature2[j].toByteArray(), 0).replace("\n".toRegex(), "") == tmpStr) {
                                    i = 1
                                }
                                j += 1
                            }
                        }
                    }
                    if (k == 0) {
                        paramAnonymousMethodHookParam.result = 0
                    }
                }
            })

        }
        if (paramLoadPackageParam.packageName == "com.android.settings") {
            XposedBridge.hookAllMethods(XposedHelpers.findClass("com.android.settings.applications.AppOpsDetails", paramLoadPackageParam.classLoader), "isPlatformSigned", object : XC_MethodHook() {
                @Throws(Throwable::class)
                override fun beforeHookedMethod(paramAnonymousMethodHookParam: XC_MethodHook.MethodHookParam) {
                    prefs.reload()
                    if (prefs.getBoolean(XpStatus.KEY_CORECRACK, false)) {
                        paramAnonymousMethodHookParam.result = false
                    }
                }
            })
        }

    }
}